# Homework 5. KNN - Handwritten digits

***Double Click here to edit this cell***

- Name: 김현동
- Student ID: 201901208
- Submission date: 2023-06-09

### We have 1797 handwritten digits of size 8x8

In [ ]:
from sklearn.datasets import load_digits
digits = load_digits()
print(digits.data.shape)

(1797, 64)


In [ ]:
digits.target.shape

(1797,)

In [ ]:
# Do not run this cell when you submit you homework
%%capture
print(digits.DESCR)

### Try with some handwritten images

In [ ]:
# Do not run this cell when you submit you homework
%%capture
%matplotlib inline

import matplotlib.pyplot as plt

k = 100
print('Target is {}'.format(digits.target[k]))
print('Data is :\n')
print(digits.data[k].reshape(8,8))
plt.gray()
plt.matshow(digits.images[k])
plt.show()

### 주의: `train_test_split`의 `random_state`를 0이 아닌 다른 값으로 바꾸거나, 다시 데이터셋을 split하면 0점 처리됨

- you must set "random_state=0"" in "train_test_split"
- don't modify the random number seed (random_state)
- Split the data into training dataset (0.67 of the total data), test dataset(0.33 of the total data)

In [ ]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(digits.data, digits.target, test_size=0.33, \
                                                    random_state=0)

# Problem 1 (10 pts)

- We want to classify handwritten digits using **1 nearest neighbor classifier**

## STEP 1

- import whatever you need
- Use sklearn 1 nearest neighbor classifier to classify digits
- Print confusion matrix, classification report, accuracy

In [ ]:
# YOUR CODE HERE
from sklearn.neighbors import KNeighborsClassifier
from sklearn.metrics import confusion_matrix,classification_report,accuracy_score

knn = KNeighborsClassifier(n_neighbors=1)
knn.fit(X_train,y_train)
y_pred = knn.predict(X_test)

cm = confusion_matrix(y_test,y_pred)
cr = classification_report(y_test,y_pred)
accuracy = accuracy_score(y_test,y_pred)

print('confusion matrix : \n')
print(cm)
print('\n\nclassification_report : \n')
print(cr)
print('Average accuracy = {:.2f}%'.format(accuracy*100))

confusion matrix : 

[[49  0  0  0  0  0  0  0  0  0]
 [ 0 61  0  0  0  0  0  0  0  0]
 [ 0  0 62  0  0  0  0  0  0  0]
 [ 0  0  0 55  0  0  0  0  0  0]
 [ 0  0  0  0 50  0  0  0  0  0]
 [ 0  0  0  0  0 63  1  0  0  1]
 [ 0  0  0  0  0  0 67  0  0  0]
 [ 0  0  0  0  0  0  0 56  0  0]
 [ 0  2  0  1  0  0  0  0 64  0]
 [ 0  0  0  3  0  1  0  0  0 58]]


classification_report : 

              precision    recall  f1-score   support

           0       1.00      1.00      1.00        49
           1       0.97      1.00      0.98        61
           2       1.00      1.00      1.00        62
           3       0.93      1.00      0.96        55
           4       1.00      1.00      1.00        50
           5       0.98      0.97      0.98        65
           6       0.99      1.00      0.99        67
           7       1.00      1.00      1.00        56
           8       1.00      0.96      0.98        67
           9       0.98      0.94      0.96        62

    accuracy            

My result:
```
confusion matrix :

[[49  0  0  0  0  0  0  0  0  0]
 [ 0 61  0  0  0  0  0  0  0  0]
 [ 0  0 62  0  0  0  0  0  0  0]
 [ 0  0  0 55  0  0  0  0  0  0]
 [ 0  0  0  0 50  0  0  0  0  0]
 [ 0  0  0  0  0 63  1  0  0  1]
 [ 0  0  0  0  0  0 67  0  0  0]
 [ 0  0  0  0  0  0  0 56  0  0]
 [ 0  2  0  1  0  0  0  0 64  0]
 [ 0  0  0  3  0  1  0  0  0 58]]


classification_report :

              precision    recall  f1-score   support

           0       1.00      1.00      1.00        49
           1       0.97      1.00      0.98        61
           2       1.00      1.00      1.00        62
           3       0.93      1.00      0.96        55
           4       1.00      1.00      1.00        50
           5       0.98      0.97      0.98        65
           6       0.99      1.00      0.99        67
           7       1.00      1.00      1.00        56
           8       1.00      0.96      0.98        67
           9       0.98      0.94      0.96        62

    accuracy                           0.98       594
   macro avg       0.99      0.99      0.99       594
weighted avg       0.99      0.98      0.98       594

Average accuracy = 98.48%
```

# Problem 2 (30 pts) - Choose k

- We want to classify handwritten digits using k nearest neighbor classifier
- Use dimensionality reduction technique like PCA or TSNE if necessary
- If you choose any hyperparameters and you did any transformation in your data, justify your choice and transformation.
- The justification must be code, data, or plotted images, **not your guess!!!**
- This must be a longer code.
- To justify your choice of hyperparameters, use visualization or plotting whenever possible.
- You may add as many cells as you want.
- To add a cell, insert->insert cell or click '+' button above in the notebook.
- you must set "random_state=0"" in "train_test_split", "TSNE" or on any functions based on random numbers

### This will be evaluated based on average accuracy

In [ ]:
# ADD YOUR CELL
import numpy as np
from sklearn.manifold import TSNE
from sklearn.decomposition import PCA

for i in range(1,10,3):
  pca = PCA(n_components=i, random_state=0)
  X_train_pca = pca.fit_transform(X_train)
  X_test_pca = pca.transform(X_test)

  knn = KNeighborsClassifier(n_neighbors=4)
  knn.fit(X_train_pca,y_train)

  y_pred = knn.predict(X_test_pca)

  accuracy = accuracy_score(y_test,y_pred)

  print(f"n={i}일때의 수치\n")
  print('Average Accuracy = {:.2f}%\n'.format(accuracy * 100))

n=1일때의 수치

Average Accuracy = 29.46%

n=4일때의 수치

Average Accuracy = 85.19%

n=7일때의 수치

Average Accuracy = 96.30%



In [ ]:
# ADD YOUR CELL
import numpy as np
from sklearn.manifold import TSNE
from sklearn.decomposition import PCA

pca = PCA(n_components=7, random_state=0)
X_train_pca = pca.fit_transform(X_train)
X_test_pca = pca.transform(X_test)

for i in range(1,10,3):

  knn = KNeighborsClassifier(n_neighbors=i)
  knn.fit(X_train_pca,y_train)

  y_pred = knn.predict(X_test_pca)

  accuracy = accuracy_score(y_test,y_pred)
  cr = classification_report(y_test, y_pred)

  print(f"k={i}일때의 수치\n")
  print('\nClassification Report:\n', cr)
  print('Average Accuracy = {:.2f}%\n'.format(accuracy * 100))

k=1일때의 수치


Classification Report:
               precision    recall  f1-score   support

           0       1.00      1.00      1.00        49
           1       0.98      0.95      0.97        61
           2       0.98      0.98      0.98        62
           3       0.88      0.96      0.92        55
           4       0.96      1.00      0.98        50
           5       0.95      0.95      0.95        65
           6       0.96      1.00      0.98        67
           7       0.96      0.98      0.97        56
           8       0.97      0.90      0.93        67
           9       0.95      0.89      0.92        62

    accuracy                           0.96       594
   macro avg       0.96      0.96      0.96       594
weighted avg       0.96      0.96      0.96       594

Average Accuracy = 95.96%

k=4일때의 수치


Classification Report:
               precision    recall  f1-score   support

           0       1.00      1.00      1.00        49
           1       0.95      0.98

(To edit, double click this cell)
### Write your justicaition of your experiment
특정 범위의 k 값에 대해 실험을 수행하여 어떤 k 값들이 좋은 성능을 나타내는지 확인했습니다. 좋은 성능의 기준은 Average Accuracy의 값이 높은 것입니다. 이를 위해 k값의 범위를 큰 값에서부터 작은 값까지 조정했습니다.

precision , recall , f1-score을 나타낸 Classification Report를 보고 k별 성능차이를 한눈에 파악하기 어려웠습니다. 따라서 전체 샘플을 대상으로 한 정확도 accuracy가 높을수록 성능이 좋다는 판단기준을 세웠습니다.

k=1일때 95.96%으로 시작해서 k=10까지 증가함에 따라 일반적으로 94~95%의 성능을 보여줍니다. k=4 일때 96.30%로 가장 높은 성능을 보여줍니다.

또한 n_neighbors의 k값 test하기 전, 차원축소 통해 가장 높은 성능을 보이는 n 값을 찾았습니다. 값이 작을수록 더 큰 정보 손실이 발생하지만 계산 비용이 감소하며, 값이 클수록 더 많은 정보를 보존하고 계산 비용이 증가합니다. 현재 데이터셋에서는 n=7일때 가장 높은 효율을 가진다는 것을 실험(코드 실행)을 통해 검증했습니다.

(제출은 3개의 샘플만을 추출하여 보여줍니다.)

(To edit, double click this cell)

### Write your final accuracy in Problem 2:  (just numbers here)

96.30

# Problem 3 (30 pts) - Other Classifiers

- You may use the following classification technique in sklearn:
  - Logistic Regression
  - Decision Trees
  - Support Vector Machines (SVM)
  - Naive Bayes
  - Perceptron
- We want to classify handwritten digits using a classifier in sklearn
- Use any feature selection technique
- If you choose any hyperparameters and you did any transformation in your data, justify your choice and transformation.
- The justification must be code, data, or plotted images, **not your guess!!!**
- This must be a longer code.
- To justify your choice of hyperparameters, use visualization or plotting whenever possible.
- You may add as many cells as you want.
- To add a cell, insert->insert cell or click '+' button above in the notebook.
- you must set "random_state=0"" in "train_test_split", "TSNE" or on any functions based on random numbers

In [ ]:
# ADD YOUR CELL
from sklearn.feature_selection import RFE
from sklearn.linear_model import LogisticRegression
from sklearn.tree import DecisionTreeClassifier
from sklearn.svm import SVC
from sklearn.naive_bayes import GaussianNB
from sklearn.linear_model import Perceptron

# Feature selection using RFE
estimator = LogisticRegression(solver='liblinear', random_state=0)
selector = RFE(estimator, n_features_to_select=30, step=1)
X_train_selected = selector.fit_transform(X_train, y_train)
X_test_selected = selector.transform(X_test)

# Logistic Regression classifier
lr = LogisticRegression(solver='liblinear', random_state=0)
lr.fit(X_train_selected, y_train)
y_pred_lr = lr.predict(X_test_selected)
accuracy_lr = accuracy_score(y_test, y_pred_lr)
cm_lr = confusion_matrix(y_test, y_pred_lr)

# Print the evaluation metrics for each classifier
print('Logistic Regression Accuracy:', accuracy_lr*100)
print('Logistic Regression Confusion Matrix:\n', cm_lr)

Logistic Regression Accuracy: 93.43434343434343
Logistic Regression Confusion Matrix:
 [[49  0  0  0  0  0  0  0  0  0]
 [ 0 54  1  0  0  0  2  0  3  1]
 [ 0  3 58  1  0  0  0  0  0  0]
 [ 0  0  1 53  0  0  0  0  1  0]
 [ 0  0  0  0 50  0  0  0  0  0]
 [ 0  0  0  0  0 63  1  0  0  1]
 [ 0  1  0  0  0  0 66  0  0  0]
 [ 0  0  0  0  2  1  0 52  0  1]
 [ 0  7  1  3  1  0  0  0 54  1]
 [ 0  0  0  2  0  1  0  1  2 56]]


In [ ]:
# Decision Tree classifier
dt = DecisionTreeClassifier(random_state=0)
dt.fit(X_train_selected, y_train)
y_pred_dt = dt.predict(X_test_selected)
accuracy_dt = accuracy_score(y_test, y_pred_dt)
cm_dt = confusion_matrix(y_test, y_pred_dt)

print('Decision Tree Accuracy:', accuracy_dt*100)
print('Decision Tree Confusion Matrix:\n', cm_dt)

Decision Tree Accuracy: 84.34343434343434
Decision Tree Confusion Matrix:
 [[46  0  0  0  1  0  0  0  0  2]
 [ 0 51  2  1  2  0  0  0  2  3]
 [ 1  6 46  1  0  1  1  2  2  2]
 [ 0  2  0 48  0  0  1  1  2  1]
 [ 1  0  0  1 45  0  0  2  1  0]
 [ 0  0  1  1  2 56  1  1  2  1]
 [ 0  0  0  0  1  1 65  0  0  0]
 [ 0  0  1  0  0  0  0 52  1  2]
 [ 1  4  0  7  1  2  0  1 42  9]
 [ 0  1  2  3  0  3  0  2  1 50]]


In [ ]:
# Support Vector Machines (SVM) classifier
svm = SVC(kernel='linear', random_state=0)
svm.fit(X_train_selected, y_train)
y_pred_svm = svm.predict(X_test_selected)
accuracy_svm = accuracy_score(y_test, y_pred_svm)
cm_svm = confusion_matrix(y_test, y_pred_svm)

print('SVM Accuracy:', accuracy_svm*100)
print('SVM Confusion Matrix:\n', cm_svm)

SVM Accuracy: 95.1178451178451
SVM Confusion Matrix:
 [[49  0  0  0  0  0  0  0  0  0]
 [ 0 57  0  0  0  0  1  0  3  0]
 [ 0  2 59  1  0  0  0  0  0  0]
 [ 0  0  2 53  0  0  0  0  0  0]
 [ 0  0  0  0 50  0  0  0  0  0]
 [ 0  0  0  1  0 62  1  0  0  1]
 [ 0  1  0  0  0  0 66  0  0  0]
 [ 0  0  0  0  1  1  0 54  0  0]
 [ 0  5  1  0  1  1  0  0 57  2]
 [ 0  0  0  3  0  1  0  0  0 58]]


In [ ]:
# Naive Bayes classifier
nb = GaussianNB()
nb.fit(X_train_selected, y_train)
y_pred_nb = nb.predict(X_test_selected)
accuracy_nb = accuracy_score(y_test, y_pred_nb)
cm_nb = confusion_matrix(y_test, y_pred_nb)

print('Naive Bayes Accuracy:', accuracy_nb*100)
print('Naive Bayes Confusion Matrix:\n', cm_nb)

Naive Bayes Accuracy: 82.15488215488216
Naive Bayes Confusion Matrix:
 [[49  0  0  0  0  0  0  0  0  0]
 [ 0 40  4  0  5  1  0  3  6  2]
 [ 0  3 49  0  0  0  0  0 10  0]
 [ 0  0  1 40  0  0  0  2  7  5]
 [ 0  0  0  0 42  0  0  8  0  0]
 [ 0  0  0  0  1 53  1  9  0  1]
 [ 0  1  0  0  0  1 65  0  0  0]
 [ 0  0  0  0  2  0  0 54  0  0]
 [ 0  3  0  3  3  0  0  7 51  0]
 [ 0  0  0  5  1  0  0  5  6 45]]


In [ ]:
# Perceptron classifier
perceptron = Perceptron(random_state=0)
perceptron.fit(X_train_selected, y_train)
y_pred_perceptron = perceptron.predict(X_test_selected)
accuracy_perceptron = accuracy_score(y_test, y_pred_perceptron)
cm_perceptron = confusion_matrix(y_test, y_pred_perceptron)

print('Perceptron Accuracy:', accuracy_perceptron*100)
print('Perceptron Confusion Matrix:\n', cm_perceptron)

Perceptron Accuracy: 90.23569023569024
Perceptron Confusion Matrix:
 [[48  0  0  0  0  0  1  0  0  0]
 [ 0 60  0  0  0  0  0  0  1  0]
 [ 0  7 55  0  0  0  0  0  0  0]
 [ 0  0  1 52  0  0  0  0  1  1]
 [ 0  2  0  0 46  0  0  2  0  0]
 [ 0  1  0  0  0 62  2  0  0  0]
 [ 0  1  0  0  0  0 66  0  0  0]
 [ 0  0  0  0  1  1  0 54  0  0]
 [ 0 12  3  2  0  0  1  2 47  0]
 [ 0  6  0  2  0  1  0  2  5 46]]


(To edit, double click this cell)
### Write your justicaition of your experiment
Logistic Regression , Decision Trees , Support Vector Machines (SVM) , Naive Bayes , Perceptron

다섯가지 분류 모델을 사용해서 모델별 성능을 테스트를 진행했습니다. 위에서 코드 실행 결과 (Accuracy , Confusion Matrix)를 확인할 수 있습니다.

Confusion Matrix를 통해 각 모델이 얼마나 정확한 예측을 했는지, 오분류 됐는지 확인 가능합니다. 비교적 정확한 예측을 진행한 모델은 Logistic Regression , SVM  두가지 입니다. 높은 예측률과 적은 오분류를 보이는만큼 93퍼센트 이상의 높은 정확도를 보입니다. 그 중에서도 SVM에는 95.11%의 가장 높은 정확도를 가짐을 확인 가능합니다.

반면에 Decision Trees , Navie Bayes, Perceptron 오분류 샘플이 많이 보이며, 비교적 낮은 성능을 보여줍니다.

5가지 모델은 지도학습 모델로, 정답인 label을 포함한 데이터를 학습한다는 점에서 비슷한 성능을 보일 것으로 생각했습니다. 하지만 각 모델별로 큰 성능 차이를 보임을 확인했습니다. 0.1%라는 작은 값이라도 AI 예측에서는 오류를 몇배 줄였다고 판단 가능한만큼 그 차이는 크다고 할 수 있습니다.

각 모델은 내부적으로 다른 가정과 알고리즘을 사용하여 데이터를 학습하고 예측합니다. 데이터셋과 문제의 특성에 따라 역전된 결과를 보여줄 수 있습니다. 이러한 결과는 데이터셋에 따른 학습 모델 선택이 중요함을 알게 해줬습니다.


------------------------------------------------------------------------------


로지스틱 회귀 : 입력 데이터의 선형 결합에 로지스틱 함수를 적용하는데, 이 때 각 입력 변수를 그대로 사용할 수도 있고, 다항식이나 상호작용 항 등의 특징을 추가로 생성할 수도 있습니다.

의사결정나무 : 특징을 추출하는 대신 데이터를 구분하는 결정 규칙(조건)을 학습합니다. 데이터를 분할하는 기준에 따라 특징이 결정되며, 이는 특징의 중요도나 유용성을 나타내는 역할을 합니다.

SVM : 입력 데이터를 고차원 특징 공간으로 매핑하는 커널 함수를 사용하여 데이터를 분류합니다. 이러한 커널 함수를 통해 데이터를 분리하는 최적의 경계를 찾게 되며, 이러한 경계에 따라 특징이 결정됩니다.

나이브 베이즈 : 데이터의 특징 간의 독립성 가정을 기반으로 학습되며, 입력 데이터의 각 특징이 주어진 클래스에 대해 조건부로 독립적으로 발생한다고 가정합니다. 이를 통해 특징의 확률 분포를 추정하고 예측을 수행합니다.

퍼셉트론 : 입력 데이터와 가중치의 선형 결합에 활성화 함수를 적용하여 출력을 계산합니다. 이때 가중치는 학습을 통해 조정되며, 각 입력 변수와 가중치의 조합에 따라 특징이 결정됩니다.



(To edit, double click this cell)

### Write your final best accuracy in Problem 3: (just numbers here)
95.11

## State-of-the-Art of MNIST classification: 99.79%

- https://rodrigob.github.io/are_we_there_yet/build/classification_datasets_results.html
- MNIST dataset is much bigger dataset: 60,000 training images and 10,000 testing images

## Ethics:
If you cheat, you will get negatgive of the total points.
If the homework total is 22 and you cheat, you get -22.

## What to submit

- Run **all cells** after restarting the kernel
- Goto "File -> Print Preview"
- Print the page as pdf
- Pdf file name must be in a form of: homework_5_홍길동_202300001.pdf
- Submit the pdf file in google classroom
- No late homeworks will be accepted
- Your homework will be graded on the basis of correctness and programming skills